src/tenant_tokens.rs

```diff
@@ -0,0 +1,115 @@
+use crate::{
+    errors::* 
+};
+use serde::{Serialize, Deserialize};
+use jsonwebtoken::{encode, Header, EncodingKey};
+use time::{OffsetDateTime};
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")] 
+struct TenantTokenClaim {
+    api_key_prefix: String,
+    search_rules: serde_json::Value,
+    #[serde(with = "time::serde::timestamp::option")]
+    exp: Option<OffsetDateTime>,
+}
+
+pub fn generate_tenant_token(search_rules: serde_json::Value, api_key: &str, expires_at: Option<OffsetDateTime>) -> Result<String, Error> {
+    if api_key.len() < 8 {
+        return Err(Error::TenantTokensInvalidApiKey)
+    }
+
+    if expires_at.is_some() && OffsetDateTime::now_utc() > expires_at.unwrap() {
+        return Err(Error::TenantTokensExpiredSignature)
+    }
+
+    let key_prefix = &api_key[..8];
```

Comment:
Here you're creating a new `&str` going from the first byte to the eighth byte.
If the `api_key` can contain utf-8 this can panic.
```suggestion
    // Unwrap is safe since we ensured before that we had at least 8 characters in the api_key
    let key_prefix = &api_key[..api_key.char_indices().nth(8).unwrap().0]];
```

Reply:
Oh, I just noticed you were doing a `.to_string` right below on the `key_prefix`.
Since you're going to allocate a string whatever happens you could write this simpler version;
```suggestion
    let key_prefix = &api_key.chars().take(8).collect();
```
And remove the `.to_string()` below.